ref(ui): Add performance event view hook

k-fish opened the pull request:
### Summary
This adds a hook to avoid having to constantly drill event view, especially for our generic discover query specializations. Essentially if any specialization now uses the 'DiscoverQueryPropsWithContext' type you no longer have to provide eventView and orgSlug, which should reduce a lot of boilerplate throughout performance. It's not on by default since there is likely some undefined checks and tests that might need to be rewritten (to use a provider) if you choose to use this.

Other:
- Giving a context wrapper to always check if a context is defined, we'll see if it's helpful, and if so we can raise this up to be used everywhere in the UI or talk about it during TSC.
- This will currently allow `usePerformanceEventView` to be used under the main performance content page and vital details, we can add it to the rest in subsequent PR's.

k-fish committed:
ref(ui): Add performance event view hook

This adds a hook to avoid having to constantly drill event view, especially for our generic discover query specializations. Essentially if any specialization now uses the 'DiscoverQueryPropsWithContext' type you no longer have to provide eventView and orgSlug, which should reduce a lot of boilerplate throughout performance. It's not on by default since there is likely some undefined checks and tests that might need to be rewritten (to use a provider) if you choose to use this.

Other:
- Giving a context wrapper to always check if a context is defined, we'll see if it's helpful, and if so we can raise this up to be used everywhere in the UI or talk about it during TSC.

Zylphrex commented on static/app/views/performance/contexts/performanceEventViewContext.tsx:
```diff
@@ -0,0 +1,22 @@
+
+// Provides a readonly event view. Also omits anything that isn't currently read-only, although in the future we should switch the code in EventView instead.
+// If you need mutability, use the mutable version.
+const usePerformanceEventView: () => ImmutableEventView = () =>
+  _usePerformanceEventView().eventView;
+export {PerformanceEventViewProvider, usePerformanceEventView};
+
+export function useMutablePerformanceEventView() {
+  return usePerformanceEventView().clone();
+}
```
Nit: this just looks inconsistent with the mix of arrow functions, normal functions, and the different syntaxes for exports.

k-fish replied:
all-the-things.jpg

Zylphrex commented on static/app/views/performance/contexts/utils.tsx:
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
```
Should this file exist within `static/app/utils/...` instead?

k-fish replied:
Definitely, good call.

k-fish committed:
Move context folder to utils

calvinlough commented on static/app/utils/discover/genericDiscoverQuery.tsx:
```diff
@@ -228,6 +244,19 @@ class GenericDiscoverQuery<T, P> extends React.Component<Props<T, P>, State<T>>
 }
 
+// Shim to allow us to use generic discover query or any specialization with or without passing org slug or eventview, which are now contexts.
+// This will help keep tests working and we can remove extra uses of context-provided props and update tests as we go.
```
Good call

k-fish replied:
Thanks.

dashed replied:
🤔

k-fish committed:
ref(ui): Add performance event view hook

This adds a hook to avoid having to constantly drill event view, especially for our generic discover query specializations. Essentially if any specialization now uses the 'DiscoverQueryPropsWithContext' type you no longer have to provide eventView and orgSlug, which should reduce a lot of boilerplate throughout performance. It's not on by default since there is likely some undefined checks and tests that might need to be rewritten (to use a provider) if you choose to use this.

Other:
- Giving a context wrapper to always check if a context is defined, we'll see if it's helpful, and if so we can raise this up to be used everywhere in the UI or talk about it during TSC.

k-fish committed:
Add event view provider to pageLayout for all txn summary pages as well

k-fish committed:
Merge branch 'ref/add-use-performance-event-view-hook' of github.com:getsentry/sentry into ref/add-use-performance-event-view-hook

k-fish committed:
Merge remote-tracking branch 'origin/master' into ref/add-use-performance-event-view-hook
